Next.js Starter Kit Guide (beta) (#583)
liveblocks/liveblocks: docs/pages/guides/nextjs-starter-kit.mdx

implement it.

#### Getting the secrets

1. Go to [Developer Settings](https://github.com/settings/apps) on GitHub and
   click "New GitHub App"
2. Enter an app name (e.g. `Liveblocks Starter Kit (dev)`). You'll need a new
   app for each environment, so it's helpful to place "dev" in the name.
3. Add a homepage URL—this isn't important now, so a placeholder will do.
4. Find the "Callback URL" input just below, and add your local development URL
   (e.g. `http://localhost:3000`)
5. Look for the "Webhook" section and make sure to uncheck "Active".
6. Use the remaining default settings and press "Create GitHub App"
7. On the next page under "Client secrets", press "Generate a new client
   secret".
8. Copy this secret into the `/.env.local` file as `GITHUB_SECRET`
9. Go back to the previous page and find the "Client ID" near the top. Copy this
   into your `.env.local` file as `GITHUB_ID`

Almost there! `.env.local` should now contain these two lines, along with anything previously there:

```
GITHUB_SECRET=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
GITHUB_ID=XXXXXXXXXXXXXXXXXXXX
```

GitHub authentication is now set up!

</details>

> 

Author: ```suggestion

<details>
  <summary>How to set up Auth0 authentication</summary>
  
  ### GitHub authentication
  
  It's possible to allow users to sign up to your app using GitHub. This is how to
  implement it.

  1. Go to your [Auth 0 Dashboard](https://manage.auth0.com/dashboard) and click "Create Application"
  2. Enter an app name (e.g. `Liveblocks Starter Kit (dev)`). You'll need a new app for each environment, so it's helpful to place "dev" in the name.
  3. Next, select "Single Page Web Applications", and press "Create"
  4. Copy your "Client ID" from the top of the page, and place it within `.env.local` as `AUTH0_CLIENT_ID`
  5. Click the "Settings" tab—we'll be making a number of changes here.
  6. Find the "Client Secret" input field, and copy the value into `.env.local` as `AUTH0_CLIENT_SECRET`
  7. Copy your "Domain" from the input field, add "https://" to the start, and place it within `.env.local` as `AUTH0_ISSUER_BASE_URL`
  8. Add the following to the "Allowed Callback URLs" textarea: `http://localhost:3000/api/auth/callback/auth0`
  
  9. Add the following to the "Allowed Logout URLs" textarea: `http://localhost:3000`
  
  10. Scroll to the bottom and press "Save changes"
  
`.env.local` should now contain these three lines, along with anything previously there:

\```
AUTH0_CLIENT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
AUTH0_CLIENT_SECRET=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
AUTH0_ISSUER_BASE_URL=https://XXXXXXXXXXXXXXXXXX.com
\```
  
  Auth0 authentication is now set up! Next, add yourself a user to test out your authentication.
</details>

```

---
meta:
  title: "Next.js Starter Kit"
  parentTitle: "Guides"
  description: "Get started with the Liveblocks Next.js Starter Kit"
---

## Features

- Documents dashboard with pagination, drafts, groups, auto-revalidation
- Collaborative whiteboard app with fully-featured share menu
- Access can be granted to users, groups, and the public
- Authentication compatible with GitHub, Google, Auth0, and more
- Type-safe client & server functions, and everything documented
- A great starting point for your collaborative application

## Liveblocks

This starter kit uses [Liveblocks](https://liveblocks.io]) to power its
collaboration. Here's how to get it running.

> 1. Create an account at [liveblocks.io](https://liveblocks.io])

Reviewer: should we mention creating an application with liveblocks package installed

Reviewer: or i guess will this get updated once its packaged?

Author: Ah, yes! 🙂 We should mention running the installer first. It'll probably look like this (it hasn't been published yet):
```
npx create-liveblocks-app --next
```

2. Navigate to the [dashboard](https://liveblocks.io/dashboard) and click "Add
   project..."
3. Create a new project with "Development" as the environment.
4. Go to "API keys" on the side menu, reveal your secret key, and copy it

> 5. Create a file called `.env.local` alongside `package.json` in your project.

Author: ```suggestion
5. Go to your downloaded starter kit, and enter the file called `.env.local` alongside `package.json` in your project.
```
The installer will now create this file for you and place in the names of the required env variables (but you have to fill in the values). There will also be a randomly generated key for NextAuth. Example:

```
NEXTAUTH_SECRET_KEY=v09ds7f09dsa7g9sadg7a9sg7as96gasg
LIVEBLOCKS_SECRET_KEY=
GITHUB_CLIENT_ID=
GITHUB_SECRET_ID=
```

6. Place your secret key into the file as `LIVEBLOCKS_SECRET_KEY`.

> Liveblocks will now work! Your `.env.local` file should look similar to this:

Author: ```suggestion
Liveblocks will now work! Your `.env.local` file should contain a line similar to this:
```

add real providers, such as GitHub, Google, Auth0.

<details>
  <summary>How to set up GitHub authentication</summary>

### GitHub authentication

It's possible to allow users to sign up to your app using GitHub. This is how to
implement it.

#### Getting the secrets

1. Go to [Developer Settings](https://github.com/settings/apps) on GitHub and
   click "New GitHub App"
2. Enter an app name (e.g. `Liveblocks Starter Kit (dev)`). You'll need a new
   app for each environment, so it's helpful to place "dev" in the name.
3. Add a homepage URL—this isn't important now, so a placeholder will do.
4. Find the "Callback URL" input just below, and add your local development URL
   (e.g. `http://localhost:3000`)
5. Look for the "Webhook" section and make sure to uncheck "Active".
6. Use the remaining default settings and press "Create GitHub App"
7. On the next page under "Client secrets", press "Generate a new client
   secret".
8. Copy this secret into the `/.env.local` file as `GITHUB_SECRET`
9. Go back to the previous page and find the "Client ID" near the top. Copy this
   into your `.env.local` file as `GITHUB_ID`

> Almost there! `.env.local` should now look similar to this:

Author: ```suggestion
Almost there! `.env.local` should now contain these two lines, along with anything previously there:
```

```

> LIVEBLOCKS_SECRET_KEY=sk_dev_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
> GITHUB_SECRET=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
> GITHUB_ID=XXXXXXXXXXXXXXXXXXXX

Author: ```suggestion
GITHUB_SECRET=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
GITHUB_ID=XXXXXXXXXXXXXXXXXXXX
```

```

> #### Add the provider
> 
> Next, go to `/pages/api/auth/[...nextAuth].ts` in your project, and add a GitHub
> provider:
> 
> ```ts
> import GithubProvider from "next-auth/providers/github";
> 
> export const authOptions = {
>   // ...
>   providers: {
>     GithubProvider({
>       clientId: process.env.GITHUB_CLIENT_ID as string,
>       clientSecret: process.env.GITHUB_CLIENT_SECRET as string,
>     }),
>   }
> };
> ```
> 
> And remove `CredentialsProvider` from the same object, as this was only necessary for the demo authentication system:
> ```ts
> export const authOptions = {
>   // ...
>   providers: {
>     // Delete from here
>     CredentialsProvider({
>       // ...
>     }),
>     // Until here
>   }
> };
> ```

Author: ```suggestion
```

The installer will now do this for you, so we can remove this section. Though we could consider keeping it in case people don't use the installer? What do you think? @taylorpwhipp
